src/gc-alloc-profiler.cpp

```diff
@@ -0,0 +1,517 @@
+    } else {
+        return false;
+    }
+}
+
+bool starts_with(string const &full_string, string const &beginning) {
+    if (full_string.length() >= beginning.length()) {
+        return (0 == full_string.compare(0, beginning.length(), beginning));
+    } else {
+        return false;
+    }
+}
+
+string _type_as_string(jl_datatype_t *type) {
+    if ((uintptr_t)type < 4096U) {
+        return "<corrupt>";
+    } else if (type == (jl_datatype_t*)jl_buff_tag) {
+        return "<buffer>";
+    } else if (type == (jl_datatype_t*)jl_malloc_tag) {
+        return "<malloc>";
+    } else if (type == jl_string_type) {
+        return "<string>";
+    } else if (type == jl_symbol_type) {
+        return "<symbol>";
+    } else if (jl_is_datatype(type)) {
+        ios_t str_;
+        ios_mem(&str_, 10024);
+        JL_STREAM* str = (JL_STREAM*)&str_;
+
+        jl_static_show(str, (jl_value_t*)type);
+
+        string type_str = string((const char*)str_.buf, str_.size);
+        ios_close(&str_);
+
+        return type_str;
+    } else {
+        return "<missing>";
+    }
+}
+
+// === stack stuff ===
+
+vector<StackFrame> get_julia_frames(jl_bt_element_t *bt_entry) {
+    vector<StackFrame> ret;
+
+    size_t ip = jl_bt_entry_header(bt_entry);
+    jl_value_t *code = jl_bt_entry_jlvalue(bt_entry, 0);
+    if (jl_is_method_instance(code)) {
+        // When interpreting a method instance, need to unwrap to find the code info
+        code = ((jl_method_instance_t*)code)->uninferred;
+    }
+    if (jl_is_code_info(code)) {
+        jl_code_info_t *src = (jl_code_info_t*)code;
+        // See also the debug info handling in codegen.cpp.
+        // NB: debuginfoloc is 1-based!
+        intptr_t debuginfoloc = ((int32_t*)jl_array_data(src->codelocs))[ip];
+        while (debuginfoloc != 0) {
+            jl_line_info_node_t *locinfo = (jl_line_info_node_t*)
+                jl_array_ptr_ref(src->linetable, debuginfoloc - 1);
+            assert(jl_typeis(locinfo, jl_lineinfonode_type));
+            const char *func_name = "Unknown";
+            jl_value_t *method = locinfo->method;
+            if (jl_is_method_instance(method))
+                method = ((jl_method_instance_t*)method)->def.value;
+            if (jl_is_method(method))
+                method = (jl_value_t*)((jl_method_t*)method)->name;
+            if (jl_is_symbol(method))
+                func_name = jl_symbol_name((jl_sym_t*)method);
+
+            ret.push_back(StackFrame{
+                func_name,
+                jl_symbol_name(locinfo->file),
+                locinfo->line,
+            });
+            
+            debuginfoloc = locinfo->inlined_at;
+        }
+    }
+    else {
+        // If we're using this function something bad has already happened;
+        // be a bit defensive to avoid crashing while reporting the crash.
+        jl_safe_printf("No code info - unknown interpreter state!\n");
+    }
+    return ret;
+}
```

Comment:
I'm getting a segfault during `GC.stop_and_write_alloc_profile("gc-profile.json")`:
```julia
No code info - unknown interpreter state!

signal (11): Segmentation fault: 11
in expression starting at REPL[6]:1
__cxx_atomic_load<_jl_value_t *> at /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/atomic:970 [inlined]
load at /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/atomic:1487 [inlined]
atomic_load_explicit<_jl_value_t *> at /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/atomic:1828 [inlined]
jl_atomic_load_explicit<_jl_value_t *> at /Users/nathandaly/src/julia/src/./julia_atomics.h:93 [inlined]
jl_array_ptr_ref at /Users/nathandaly/src/julia/src/./julia.h:950 [inlined]
get_julia_frames at /Users/nathandaly/src/julia/src/gc-alloc-profiler.cpp:193
get_frames at /Users/nathandaly/src/julia/src/gc-alloc-profiler.cpp:266
expand_stack at /Users/nathandaly/src/julia/src/gc-alloc-profiler.cpp:299
profile_serialize at /Users/nathandaly/src/julia/src/gc-alloc-profiler.cpp:350
alloc_profile_serialize at /Users/nathandaly/src/julia/src/gc-alloc-profiler.cpp:420
jl_stop_and_write_alloc_profile at /Users/nathandaly/src/julia/src/gc-alloc-profiler.cpp:439
stop_and_write_alloc_profile at ./gcutils.jl:113 [inlined]
#1 at ./REPL[2]:3 [inlined]
#open#363 at ./io.jl:384
open at ./io.jl:382 [inlined]
stop_and_write_alloc_profile at ./REPL[2]:2 [inlined]
top-level scope at ./REPL[6]:6
jl_toplevel_eval_flex at /Users/nathandaly/src/julia/src/toplevel.c:880
jl_toplevel_eval_flex at /Users/nathandaly/src/julia/src/toplevel.c:833
jl_toplevel_eval_flex at /Users/nathandaly/src/julia/src/toplevel.c:833
ijl_toplevel_eval at /Users/nathandaly/src/julia/src/toplevel.c:898 [inlined]
ijl_toplevel_eval_in at /Users/nathandaly/src/julia/src/toplevel.c:948
```

Reply:
@maleadt does anything stand out to you here? @vilterp says he copied this code from `stackwalk.c`.

Reply:
No, but I'm not too familiar with that code.